<a href="https://colab.research.google.com/github/KumudaBG/Forest-Fire-Prediction/blob/main/Data_collection_ArcGIS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Data Extraction
We are provided with the SQL Lite data containing a dataset from NWCG from year 1992-2015. In this particular step we are trying to:-
1. Extract Data from SQL Lite database
2. Filtering out data for California
3. Grouping data based on county and seeing which county had the most fire incidents
4. This will help us to narrow down to 1 county to start our work with.

In [ ]:
# Importing data from SQL Lite to Pandas Dataframe
import sqlite3
cnx = sqlite3.connect('./FPA_FOD_20170508.sqlite')
df = pd.read_sql_query("SELECT * FROM fires", cnx)


In [ ]:
# Seeing the sample data
df.head(4)

OBJECTID  FOD_ID      FPA_ID SOURCE_SYSTEM_TYPE SOURCE_SYSTEM  \
0         1       1  FS-1418826                FED   FS-FIRESTAT   
1         2       2  FS-1418827                FED   FS-FIRESTAT   
2         3       3  FS-1418835                FED   FS-FIRESTAT   
3         4       4  FS-1418845                FED   FS-FIRESTAT   

  NWCG_REPORTING_AGENCY NWCG_REPORTING_UNIT_ID  NWCG_REPORTING_UNIT_NAME  \
0                    FS                USCAPNF    Plumas National Forest   
1                    FS                USCAENF  Eldorado National Forest   
2                    FS                USCAENF  Eldorado National Forest   
3                    FS                USCAENF  Eldorado National Forest   

  SOURCE_REPORTING_UNIT SOURCE_REPORTING_UNIT_NAME  ... FIRE_SIZE_CLASS  \
0                  0511     Plumas National Forest  ...               A   
1                  0503   Eldorado National Forest  ...               A   
2                  0503   Eldorado National Forest  ...               A   
3                  0503   Eldorado National Forest  ...               A   

    LATITUDE   LONGITUDE OWNER_CODE       OWNER_DESCR STATE COUNTY FIPS_CODE  \
0  40.036944 -121.005833        5.0              USFS    CA     63       063   
1  38.933056 -120.404444        5.0              USFS    CA     61       061   
2  38.984167 -120.735556       13.0  STATE OR PRIVATE    CA     17       017   
3  38.559167 -119.913333        5.0              USFS    CA      3       003   

   FIPS_NAME                                              Shape  
0     Plumas  b'\x00\x01\xad\x10\x00\x00\xe8d\xc2\x92_@^\xc0...  
1     Placer  b'\x00\x01\xad\x10\x00\x00T\xb6\xeej\xe2\x19^\...  
2  El Dorado  b'\x00\x01\xad\x10\x00\x00\xd0\xa5\xa0W\x13/^\...  
3     Alpine  b'\x00\x01\xad\x10\x00\x00\x94\xac\xa3\rt\xfa]...  

[4 rows x 39 columns]

In [ ]:
# Checking the infor about the data frame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1880465 entries, 0 to 1880464
Data columns (total 39 columns):
 #   Column                      Dtype  
---  ------                      -----  
 0   OBJECTID                    int64  
 1   FOD_ID                      int64  
 2   FPA_ID                      object 
 3   SOURCE_SYSTEM_TYPE          object 
 4   SOURCE_SYSTEM               object 
 5   NWCG_REPORTING_AGENCY       object 
 6   NWCG_REPORTING_UNIT_ID      object 
 7   NWCG_REPORTING_UNIT_NAME    object 
 8   SOURCE_REPORTING_UNIT       object 
 9   SOURCE_REPORTING_UNIT_NAME  object 
 10  LOCAL_FIRE_REPORT_ID        object 
 11  LOCAL_INCIDENT_ID           object 
 12  FIRE_CODE                   object 
 13  FIRE_NAME                   object 
 14  ICS_209_INCIDENT_NUMBER     object 
 15  ICS_209_NAME                object 
 16  MTBS_ID                     object 
 17  MTBS_FIRE_NAME              object 
 18  COMPLEX_NAME                object 
 19  FIRE_YEAR            

In [ ]:
df.OBJECTID.count()

1880465

We can see here that we are having 1880464 data records

**Filtering data with respect to CA**

From the below mentioned max count it is pretty evident that california has the maximum number of fire event.

### Conclusion of the Data Extraction
As we have seen from above that the state with most number of fire is california then lets pull this data out from the dataframe and get the list of all the CA records.

In [ ]:
# California Wild Fire incident dataset is stores in the file 'California_Forest_Fire_Data.csv'
dfCali = df.loc[df.STATE == "CA"]
print(dfCali.STATE.count())
dfCali.to_csv('California_Forest_Fire_Data.csv', index=False)

189550


### In this section we will try to narrow down the California wildfire data in order to figure out a county that has the most number of wildfire in all these year

In [ ]:
# Inspecting the data for California
dfCali.head()

OBJECTID  FOD_ID      FPA_ID SOURCE_SYSTEM_TYPE SOURCE_SYSTEM  \
0         1       1  FS-1418826                FED   FS-FIRESTAT   
1         2       2  FS-1418827                FED   FS-FIRESTAT   
2         3       3  FS-1418835                FED   FS-FIRESTAT   
3         4       4  FS-1418845                FED   FS-FIRESTAT   
4         5       5  FS-1418847                FED   FS-FIRESTAT   

  NWCG_REPORTING_AGENCY NWCG_REPORTING_UNIT_ID  NWCG_REPORTING_UNIT_NAME  \
0                    FS                USCAPNF    Plumas National Forest   
1                    FS                USCAENF  Eldorado National Forest   
2                    FS                USCAENF  Eldorado National Forest   
3                    FS                USCAENF  Eldorado National Forest   
4                    FS                USCAENF  Eldorado National Forest   

  SOURCE_REPORTING_UNIT SOURCE_REPORTING_UNIT_NAME  ... FIRE_SIZE_CLASS  \
0                  0511     Plumas National Forest  ...               A   
1                  0503   Eldorado National Forest  ...               A   
2                  0503   Eldorado National Forest  ...               A   
3                  0503   Eldorado National Forest  ...               A   
4                  0503   Eldorado National Forest  ...               A   

    LATITUDE   LONGITUDE OWNER_CODE       OWNER_DESCR STATE COUNTY FIPS_CODE  \
0  40.036944 -121.005833        5.0              USFS    CA     63       063   
1  38.933056 -120.404444        5.0              USFS    CA     61       061   
2  38.984167 -120.735556       13.0  STATE OR PRIVATE    CA     17       017   
3  38.559167 -119.913333        5.0              USFS    CA      3       003   
4  38.559167 -119.933056        5.0              USFS    CA      3       003   

   FIPS_NAME                                              Shape  
0     Plumas  b'\x00\x01\xad\x10\x00\x00\xe8d\xc2\x92_@^\xc0...  
1     Placer  b'\x00\x01\xad\x10\x00\x00T\xb6\xeej\xe2\x19^\...  
2  El Dorado  b'\x00\x01\xad\x10\x00\x00\xd0\xa5\xa0W\x13/^\...  
3     Alpine  b'\x00\x01\xad\x10\x00\x00\x94\xac\xa3\rt\xfa]...  
4     Alpine  b'\x00\x01\xad\x10\x00\x00@\xe3\xaa.\xb7\xfb]\...  

[5 rows x 39 columns]

In [ ]:
# Checking the various data types California Data Frame has
dfCali.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 189550 entries, 0 to 1880464
Data columns (total 39 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   OBJECTID                    189550 non-null  int64  
 1   FOD_ID                      189550 non-null  int64  
 2   FPA_ID                      189550 non-null  object 
 3   SOURCE_SYSTEM_TYPE          189550 non-null  object 
 4   SOURCE_SYSTEM               189550 non-null  object 
 5   NWCG_REPORTING_AGENCY       189550 non-null  object 
 6   NWCG_REPORTING_UNIT_ID      189550 non-null  object 
 7   NWCG_REPORTING_UNIT_NAME    189550 non-null  object 
 8   SOURCE_REPORTING_UNIT       189550 non-null  object 
 9   SOURCE_REPORTING_UNIT_NAME  189550 non-null  object 
 10  LOCAL_FIRE_REPORT_ID        61933 non-null   object 
 11  LOCAL_INCIDENT_ID           127983 non-null  object 
 12  FIRE_CODE                   55533 non-null   object 
 13  FIRE_NAME    

In [ ]:
# Checking the dataframe for null values
dfCali.isnull().sum()

OBJECTID                           0
FOD_ID                             0
FPA_ID                             0
SOURCE_SYSTEM_TYPE                 0
SOURCE_SYSTEM                      0
NWCG_REPORTING_AGENCY              0
NWCG_REPORTING_UNIT_ID             0
NWCG_REPORTING_UNIT_NAME           0
SOURCE_REPORTING_UNIT              0
SOURCE_REPORTING_UNIT_NAME         0
LOCAL_FIRE_REPORT_ID          127617
LOCAL_INCIDENT_ID              61567
FIRE_CODE                     134017
FIRE_NAME                      14993
ICS_209_INCIDENT_NUMBER       186712
ICS_209_NAME                  186712
MTBS_ID                       188413
MTBS_FIRE_NAME                188413
COMPLEX_NAME                  188623
FIRE_YEAR                          0
DISCOVERY_DATE                     0
DISCOVERY_DOY                      0
DISCOVERY_TIME                 79333
STAT_CAUSE_CODE                    0
STAT_CAUSE_DESCR                   0
CONT_DATE                      97642
CONT_DOY                       97642
C

We observe that there is no null data in latitude and longitude. All the data frames with respect to latitude and longitude are filled. We also see that there are certain counties that are missing. Lets use the Latitude and Longitude data to fill all the County Information.

### Generating County information as per Latitude and Longitude using Geopy package

In [ ]:
# importing Geopy with functools
from geopy.geocoders import ArcGIS
from geopy.extra.rate_limiter import RateLimiter
import json

# Creating a geolocator variable with dummy app name
geolocator = ArcGIS()
reverse = RateLimiter(geolocator.reverse)

# Reverse query in order to check whether its working fine or not
county = reverse(("37.3337, -121.8907"), exactly_one=True).raw
county['Subregion']

'Santa Clara County'

In [ ]:
# Method generating county name using the latitude and longitude data.
def getCountyInfo(lat_lang_str):
    try:
        return reverse((lat_lang_str), exactly_one=True).raw
    except:
        pass

In [ ]:
# county_geocode_list = []
# dfCaliCountyNull = dfCali[dfCali.COUNTY.isnull()]
# count = 0
# for index,row in dfCaliCountyNull.iterrows():
#     if count % 1000 == 0:
#         with open(str(count)+'countyData.json', 'w') as f:
#             json.dump(county_geocode_list, f)
#     latitude = repr(row['LATITUDE'])
#     longitude = repr(row['LONGITUDE'])
#     geoData = getCountyInfo(latitude+","+longitude)
#     geoData['Latitude'] = latitude;
#     geoData['Longitude'] = longitude;
#     county_geocode_list.append(geoData)
#     count = count+1
#     print(index)
dfCali['NewCountyValue'] = None

for index,row in dfCali.iterrows():
    try:
        latitude = repr(row['LATITUDE'])
        longitude = repr(row['LONGITUDE'])
        geoData = getCountyInfo(latitude+","+longitude)
        geoData['Latitude'] = latitude
        geoData['Longitude'] = longitude
        dfCali.at[index, 'NewCountyValue']=geoData['Subregion']
        print(dfCali['NewCountyValue'][index])
        print(index)
    except:
        print('Exception found')
        continue

In [ ]:
dfCali.to_csv('FinalCaliData.csv', index=False)